src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -14,105 +14,97 @@ namespace System.Buffers
     public readonly partial struct ReadOnlySequence<T>
     {
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, out ReadOnlyMemory<T> data, out SequencePosition next)
+        private bool TryGetBuffer(in SequencePosition position, out ReadOnlyMemory<T> memory, out SequencePosition next)
         {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
             next = default;
-            object startObject = start.GetObject();
-            if (startObject != null)
-            {
-                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
-            }
-            else
+            object positionObject = position.GetObject();
+            if (positionObject == null)
             {
-                type = SequenceType.Empty;
+                memory = default;
+                return false;
             }
 
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
-
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
-                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
+            int endIndex = GetIndex(_sequenceEnd);
+            //int length = GetIndex(_sequenceEnd) - positionIndex;
 
+            SequenceType type = GetSequenceType();
             if (type == SequenceType.MultiSegment)
             {
-                Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
-                ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
+                Debug.Assert(positionObject is ReadOnlySequenceSegment<T>);
+                ReadOnlySequenceSegment<T> positionSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(positionObject);
 
-                data = startSegment.Memory;
+                // Bounds check
+                #if DEBUG
```

Comment:
These checks need to be in Release as well (since we need to throw an exception that would be visible to the user if they passed in invalid input).

Reply:
Old code does not make correct Bounds checks. Do make bounds check in production code? It will slow down performance.

Reply:
Should we check in production next ? 
1.  `Debug.Assert(positionObject is ReadOnlySequenceSegment<T>);`
2.  
```
                Debug.Assert(startSegment.RunningIndex - positionSegment.RunningIndex <= GetIndex(position) - GetIndex(_sequenceStart) ||
                             GetIndex(position) - GetIndex(_sequenceEnd) <= endSegment.RunningIndex - positionSegment.RunningIndex);
```
3. `Debug.Assert(nextSegment != null);`
4. `Debug.Assert(positionObject == _sequenceEnd.GetObject())`
5. `Debug.Assert(GetIndex(_sequenceStart) <= GetIndex(position) && GetIndex(position) <= GetIndex(_sequenceEnd));`